### PR TITLE
Arm/feature/bft forks

### PR DESCRIPTION
**Pull request by arminsabouri**
This PR:
* Moved extra data header to reth primitives crate
* Moves validate sigs helpers to header ext trait
* Validate block should validate that a block signer will not sign for a fork that is deeper than 1 block!
* Configure blockchain tree engine to only allow for reorgs of 1 depth
* Started mocking out block chain tree ext for mock eth provider

### CODE REVIEW

File: crates/consensus/authority/src/pbft.rs [hunk 1]
```diff
@@ -200,6 +272,79 @@ impl<F: ToFrostManager> PbftStateMachine<F> {
...
+        // or a fork
+        if block_to_sign.parent_hash == best_block_hash {
+            return Ok(());
+        } else if best_block.header.number == 0 {
+            // Case where the best block is the genesis block
+            // This should never happen
+            return Ok(());
+        } else {
+            // Somehow we have the parent block but its not the current canon chain?
+            // This should not happen
```
Comment (scottmillner) [inline crates/consensus/authority/src/pbft.rs:386]: can we have the parent block, it is in the canonical chain, but it isn't `best_block_hash` bc it's deeper than that?

File: crates/consensus/authority/src/pbft.rs [hunk 2]
```diff
@@ -200,6 +272,79 @@ impl<F: ToFrostManager> PbftStateMachine<F> {
...
+        }
+
+        // Check if we are building on a block that is in the canonical chain
+        // or a fork
+        if block_to_sign.parent_hash == best_block_hash {
+            return Ok(());
+        } else if best_block.header.number == 0 {
+            // Case where the best block is the genesis block
+            // This should never happen
+            return Ok(());
```
Comment (arminsabouri) [inline crates/consensus/authority/src/pbft.rs:383]: TODO maybe we should error out here and not sign

File: crates/consensus/authority/src/pbft.rs (hunk 1 above)
Comment (arminsabouri) [reply crates/consensus/authority/src/pbft.rs:386]: Its possible but a non-byzantine node would never propose a block like that.

File: bin/reth/src/poa_builder.rs [hunk 3]
```diff
@@ -366,7 +369,13 @@ impl<DB: Database + DatabaseMetrics + DatabaseMetadata + 'static> NodeBuilderWit
...
 
         // configure blockchain tree
-        let tree_config = BlockchainTreeConfig::default();
+        // Rest of these block chain tree configs are defaults
+        let tree_config = BlockchainTreeConfig::new(
+            POA_MAX_REORG_DEPTH,
+            65,  /* max_blocks_in_chain */
+            256, /* num_of_additional_canonical_block_hashes */
+            200, /* max_unconnected_blocks */
+        );
```
Comment (BatiGencho) [inline bin/reth/src/poa_builder.rs:378]: Just a comment though - I remember the last 3 params did not have any effect when I was dealing a few months ago with the chain forks. Just fyi

File: crates/primitives/src/header_ext.rs [hunk 4]
```diff
@@ -130,6 +192,17 @@ impl HeaderExt for Header {
         Ok(())
     }
 
+    /// Validate timestamp
+    fn validate_timestamp(&self, current_timestamp: u64) -> Result<(), ValidateInturnError> {
+        // Time stamp should be less that or greater than by 2 seconds
```
Comment (BatiGencho) [inline crates/primitives/src/header_ext.rs:197]: In my MR: https://github.com/botanix-labs/Macbeth/pull/190/files this has been checked in a different way. We have to merge these eventually.

File: crates/consensus/authority/src/block_fetcher.rs [hunk 5]
```diff
@@ -266,6 +278,40 @@ where
                         }
                     }
 
+                    // Need to decide if we accepting a forked block or not
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_fetcher.rs:281]: Note: this is not tested yet

File: crates/consensus/authority/src/block_fetcher.rs [hunk 6]
```diff
@@ -266,6 +278,40 @@ where
...
+                    // There is a garuntee a quorum of signers will not sign an invalid fork
+                    let tip = storage.client.best_block_number().expect("best block exists");
+                    let best_block = storage
+                        .client
+                        .block_by_number(tip)
+                        .expect("best block exists")
```
Comment (BatiGencho) [inline crates/consensus/authority/src/block_fetcher.rs:287]: There is quite a few expects in here. Should we not just say that if .is_err() { continue } for each of these in the validation logic ?

File: bin/reth/src/poa_builder.rs (hunk 3 above)
Comment (arminsabouri) [reply bin/reth/src/poa_builder.rs:378]: Yeah I agree. The configs passed to the BTE largely have no affect

File: crates/consensus/authority/src/block_fetcher.rs (hunk 6 above)
Comment (arminsabouri) [reply crates/consensus/authority/src/block_fetcher.rs:287]: the three expects here are for getting the best_block / tip. The DB should always have a tip. Should we not panic if one can't be found?

Comment (BatiGencho) [reply crates/consensus/authority/src/block_fetcher.rs:287]: I agree with you. What if its the genesis ? (ignore the comment if its not relevant)

**Review by BatiGencho — APPROVED** [review]

**Comment by arminsabouri** [comment]
>I agree with you. What if its the genesis ? (ignore the comment if its not relevant) 
if you only have block 0. The tip should then just be genesis.

**Pull request merged**